### ISSUE TITLE
Process elevation

### ISSUE THREAD

**Issue by ProgerXP** [ProgerXP/Notepad2e#166]
It is a normal situation when user tries to edit and save a system file only to discover that he has no access to it. Currently he has to close the program, run it as admin, locate the file and redo the edits, which is very inconvenient.

How complex it is to add a command that will obtain admin permissions for this particular process (for its lifetime) without restarting it and losing the state? In other words, similar to how Explorer behaves when you try to change a system file.

If it's possible, then add `Elevate` command after File > Create Desktop Link:

* on XP (where UAC is unavailable) it's disabled
* if the process is already running under admin, or has been elevated - the command has a checkbox, and the command is disabled (no need to un-elevate)
* in other cases the command is enabled with a checkbox, and calling it brings up UAC prompt
* in any case, show the UAC shield icon before the command's text (source: https://msdn.microsoft.com/en-us/library/windows/desktop/dn742497(v=vs.85).aspx#guidelines)

**Comment by cshnik** [comment]
There is a very limited number of ways for obtaining administrators privileges for application.
[Running with Administrator Privileges article](https://msdn.microsoft.com/en-us/library/windows/desktop/ms717801(v=vs.85).aspx) from M$.
The only possible way to implement this feature without application restart is to authenticate the user by calling ``CredUIPromptForCredentials`` which will cause system dialog popup to request user credentials. Created user access token with ``LogonUser``-call will later be used when accessing system files.
Please confirm if this method sounds reasonable and we should proceed on this task.

**Comment by ProgerXP** [comment]
The method sounds reasonable, Control Panel works the same way. An UAC prompt is what I had in mind.

> Created user access token with LogonUser-call will later be used when accessing system files.

Am I correct that the only place in the code you'll need to change is `FileIO()` (or nearby)? How is it going to use the logon token to write the system file?

**Comment by cshnik** [comment]
> An UAC prompt is what I had in mind.

UAC prompt is diplayed when application requires admin privileges on it's startup.

> Am I correct that the only place in the code you'll need to change is `FileIO()` (or nearby)? How is it going to use the logon token to write the system file?

You are right. There will be ``ImpersonateLoggedOnUser``-call before the actual IO-operations and ``RevertToSelf``-call after them. This will limit the area where provided credentials are used.

**Comment by ProgerXP** [comment]
> UAC prompt is diplayed when application requires admin privileges on it's startup.

Not only on startup. Windows Explorer invokes it all the time, for example:

- copy or paste a system file - the dialog will have "continue" button that brings up UAC without recreating the progress dialog
- file security attributes (Properties > Security tab) has "change settings that are currently unavailable" button that similarly causes UAC prompt (this time the dialog is reopened though)

Obviously, for Notepad 2e UAC prompt shouldn't cause window recreation, it should be transparent for the user (preserving all state).

**Comment by cshnik** [comment]
Please check this article: [User Account Control](https://msdn.microsoft.com/en-us/library/windows/desktop/dn742497(v=vs.85).aspx)
As you may note, there is always ``Program name`` provided in every UAC prompt. Please use ProcessMonitor to make sure there are no actual commands/processes started when prompt accepted.

**Comment by ProgerXP** [comment]
It appears you are correct. The page suggests:

> (3.) **Separate elevated process.** Internally, a new elevated process is created to perform the task.

This is supported if I do this test:

1. Run Task Manager, note its PID
2. Click on "Show tasks from all users", accept UAC
3. Note that now taskmgr's PID is different

What confused me with Explorer is that it clearly doesn't create another explorer.exe with different privileges (and PID) but instead creates new dllhost process which isn't connected to Explorer's dialogs in Task Manager at all.

In this case I suggest we add a new switch to notepad2e.exe which will be only given when the program is being launched by UAC in elevated mode. The switch should make the program write given file data to given file path and then just exit.

This way the main (non-privileged) process keeps running and simply waits until elevated process finishes or elevation was rejected, then resumes and user can continue editing in the same old process as usual.

----------

First problem is - how to pass that data? Using temporary file is slow and not desirable. Something like shared memory would be ideal but AFAIK Windows only allows mapping files, which again requires a temp file. 

If shared memory isn't an option, then perhaps we can use IPC to transfer blocks of data from non-elevated process to the elevated one? Something like

```
// in elevated
target = fopen('\\protected\\file.txt', 'w');
while (!feof(ipc)) {
  fwrite(target, fread(ipc, 8096));
}
```

-------------

Second problem - would it require re-elevation every time I want to save the file? The article says:

> Note that tasks don't "remember" elevated states. For example, if the user navigates back and forth over an elevation entry point in a wizard, the user must elevate each time.

That would be awfully inconvenient. At the same time, we can't just switch to the elevated process because we'd lose the program's state (undo history, settings only effective for this instance, etc.).

**Comment by cshnik** [comment]
> shared memory would be ideal but AFAIK Windows only allows mapping files

Actually no, physical file is not required for file mapping: https://docs.microsoft.com/ru-ru/windows/desktop/Memory/creating-named-shared-memory
Named pipes or sockets can be also used for IPC: https://stackoverflow.com/questions/1906166/inter-process-communication-recommendation

> Second problem - would it require re-elevation every time I want to save the file?

UAC prompt is a part of the application startup procedure. Thus, re-elevation with UAC prompt will occur on every application start.
Another option would be to implement slave mode for application which means after elevated process start it should be persistently running (wait for data) until main application quit. This slave instance will be used (receive data from main application) only in "access denied"-cases. This way allow to have only a single UAC prompt confirmation for multiple file saving.

**Comment by ProgerXP** [comment]
> Another option would be to implement slave mode for application which means after elevated process start it should be persistently running (wait for data) until main application quit. 

Yes, it seems the most convenient way in this situation even if harder to implement.

I see the following algorithm when saving to a privileged file:

1. Toggle program UI to "saving" state
2. Generate some random identifier 
3. Check if a named file-less mapping exists with that name
4. If not - launch a second instance of the program's EXE in elevation, passing on the command line some "slave" switch and that random identifier, and also own PID; if elevation fails, go to step 8
5. Wait until mapping exists (second instance has initialized), going to step 8 after some timeout 
6. Write full buffer (to be saved) into that mapping and signal the object
7. Wait until signal object is reset (success - step 9), going to step 8 after some timeout 
8. Alert the user about error and cancel saving, preserving "modified" buffer state
9. Restore normal program UI

And second instance (obviously, coded entirely inside a 2e-specific source file):

1. Creates a global mapping with identifier from the command line
2. Enters a loop, checking if given PID (main program) is still present on the system
3. Inside the loop, waits_for_single_object (let's say a mutex or something else)
4. When that object signals, it means the mapping has full file content to be written
5. Opens target file, writes mapping contents there, closes the file, clears the mapping and the signal object; in case of any error goes to step 7
6. Repeats from step 2
7. On exit, dismisses the mapping

The second instance should also ensure that the mapping hasn't changed while it was being written. Ideally if Windows API provides a way to change mapping access  after creation, it should set it to read-only after step 3 and set it back to read/write after step 5. 

What are your ideas?

**Comment by cshnik** [comment]
> 4. When that object signals, it means the mapping has full file content to be written
> 5. Opens target file, writes mapping contents there, closes the file, clears the mapping and the signal object; in case of any error goes to step 7

Since file mapping (FM) is created by the elevated process (slave) there is a possibility (not 100% sure) we can directly use destination file to write data skipping steps 4-5, i.e:

SLAVE:
1. Create/open destination file on a restricted path
2. Create FM for this physical file
3. Signal to MAIN FM is ready

MAIN:
1. Open FM
2. Write data, close FM

I'll check if this scheme will work.

Additional IPC is required to transfer error code to provide feedback for the user if there is a problem occurred.

**Comment by ProgerXP** [comment]
> Additional IPC is required to transfer error code to provide feedback for the user if there is a problem occurred.

Why not send a message to PID's main window? Since it's a uni-directional interaction, main window doesn't need to interact with the slave.

> Since file mapping (FM) is created by the elevated process (slave) there is a possibility (not 100% sure) we can directly use destination file to write data skipping steps 4-5, i.e:

Yes, that would be a very nice and simpler approach.

**Comment by cshnik** [comment]
Please check test build.

**Comment by ProgerXP** [comment]
1. Elevated instance must be restarted if it's no more running (e.g. kill via task manager).
 2. Run the program, Elevate - mouse pointer will have "waiting" spinner for some time, why?
 3. lstrcpy can overflow the buffer if parameters are unexpectedly changed in the future, use lstrcpyn instead
 4. If file becomes non-readable, saving breaks: run, open file, change its permissions, Elevate - you will see several error messages and elevation will time out instead of displaying save error immediately
 5. I don't like waiting for INFINITE - once during testing I got the main process freezing on save (but can't reproduce); in all cases there must be some safety timeout (e.g. 1 minute) that will cause saving to fail (server)/exit (client), leaving buffer as "modified", and client will be restarted on next save
 6. Do you need read access on the file in client anyway?
 7. In n2e_RunElevatedInstance, is there any meaning in a while loop without sleeping?
 8. Current UAC icon isn't pretty and too large, can we use standard one? Also avoiding converting it to bitmap on runtime?
 9. As far as I see there's no check if all data was really written via the mapping or not. Since we don't directly write file data in client (it's done via the mapping), at least we should check file size in server after client reported successful save and if it's different from what we expected, display an error like `Unexpected elevated save error. File may be corrupted. Please save to another file.` and leave buffer as "modified"

---

> The second instance should also ensure that the mapping hasn't changed while it was being written. Ideally if Windows API provides a way to change mapping access after creation, it should set it to read-only after step 3 and set it back to read/write after step 5.

Can this be done?

**Comment by cshnik** [comment]
Please try [test build](https://github.com/ProgerXP/Notepad2e/raw/b40d2582d86a325354c16f71411feb419ec821fe/bin/Notepad2e.exe).

**Comment by ProgerXP** [comment]
> user is able to enable and disable elevation mode when working with restricted instance.

Good idea.

>> Elevated instance must be restarted if it's no more running (e.g. kill via task manager).

> Fixed.

Now try this: run the program, elevate, kill elevated process - the main window will trigger elevation (UAC) again - by itself, without my command - and if I cancel UAC it will even complain about failed elevation.

What I meant by my previous message was to restart only when user tries to save - not automatically of course. So on save you check, if mutex/pipe/mapping/etc. are not active and File > Elevate is checked, then you try to re-elevate.

>> Run the program, Elevate - mouse pointer will have "waiting" spinner for some time, why?

> The possible reason was: execution of the elevated instance was made by synchronous ShellExecute-call which does not respond until user accepted/canceled application start. Now this code moved to a separate thread. Please test new build.

No difference for me. You can move the call back into the thread, it will simplify the code.

>> Do you need read access on the file in client anyway?

> Fixed.

If the file is write-only (no read access) for admins, it still reports as if the file can't be written. Is it normal?

> File size is always correct due to the next points:

Is there any explicit notification by the client that the process has indeed finished writing operations (truncating the file, etc.)?

>> The second instance should also ensure that the mapping hasn't changed while it was being written. Ideally if Windows API provides a way to change mapping access after creation, it should set it to read-only after step 3 and set it back to read/write after step 5.

> FlushViewOfFile()-added to dump every buffer to disk (writes to the disk a byte range within a mapped view of a file). Is this enough?

Flushing wasn't what I had in mind, it's unnecessary and should happen automatically. 

What I meant is that there might be some chance that the server begins second writing (file save) session while client is still finishing writing the first session? This way part of mapping has new data, part has old data.

-------

New problem: open a file that's readable by your user but writeable by admin only, elevate, delete all contents, Save empty file (size = 0), now type single Space (so buffer size = 1; no need to Save now).

Issue: buffer isn't marked as modified (no `*` in the titlebar). Moreover, if you save then after reloading (F5) it's clear that the file wasn't written at all (size = 0).

This doesn't seem to occur when elevation is not used. The "trim trailing blanks" setting is unrelated to this issue.

Related problem: open a similar non-writeable non-empty file, delete all contents, remove write permissions from admin too (or make file readonly), Save - no error will be displayed, and you will notice the file on disk is unchanged (F5).

**Comment by cshnik** [comment]
> What I meant by my previous message was to restart only when user tries to save - not automatically of course. So on save you check, if mutex/pipe/mapping/etc. are not active and File > Elevate is checked, then you try to re-elevate.

Fixed.

>>> Do you need read access on the file in client anyway?

>> Fixed.

> If the file is write-only (no read access) for admins, it still reports as if the file can't be written. Is it normal?

It looks like there is problem with file access configuration. In my local tests I can properly save content to write-only file, i.e:
![image](https://user-images.githubusercontent.com/8019354/43796242-e46694ca-9aad-11e8-817b-5f658756e130.png)

> Is there any explicit notification by the client that the process has indeed finished writing operations (truncating the file, etc.)?

As we agreed elevated process does the following:
1. Open file handle to destination file
2. Create file mapping object for this file
3. Notify main process file mapping is ready
Main process works directly with IPC shared file mapping object (and controls operation results). In the end of this writing it signals elevated instance to finish file access operation.
4. Close file mapping object and file handle
5. Responds with error code if there was a problem

> Flushing wasn't what I had in mind, it's unnecessary and should happen automatically. 

> What I meant is that there might be some chance that the server begins second writing (file save) session while client is still finishing writing the first session? This way part of mapping has new data, part has old data.

1. dwShareMode set to 0 when creating file mapping to prevent concurrent operations from the foreign applications.
2. Current implementation does not allow to begin second writing session when accessing file mapping because writing of the contents is called from the main application thread (there is no way to get nested call stack or concurrent calls from different threads).

-------

> New problem: open a file that's readable by your user but writeable by admin only, elevate, delete all contents, Save empty file (size = 0), now type single Space (so buffer size = 1; no need to Save now).

> Issue: buffer isn't marked as modified (no `*` in the titlebar). Moreover, if you save then after reloading (F5) it's clear that the file wasn't written at all (size = 0).

> This doesn't seem to occur when elevation is not used. The "trim trailing blanks" setting is unrelated to this issue.

Fixed.

> Related problem: open a similar non-writeable non-empty file, delete all contents, remove write permissions from admin too (or make file readonly), Save - no error will be displayed, and you will notice the file on disk is unchanged (F5).

Fixed.

Please check [test build](https://github.com/ProgerXP/Notepad2e/blob/242748eaae6118906a9f982625e05462b8d72bac/bin/Notepad2e.exe?raw=true).

**Comment by ProgerXP** [comment]
I see you have added another pipe, now for the client. What for, which bug is it meant to fix?

**Comment by cshnik** [comment]
Different pipes are needed to implement IPC between MAIN/SLAVE instances: pipe1 is used for reading in MAIN and for writing in SLAVE, pipe2 is used for writing in MAIN and for reading in SLAVE.
Implementation of the IPC based on a single pipe (duplex mode) will require additional work and will definitely complicate the code.

**Comment by ProgerXP** [comment]
What I mean is that it used to work with just one pipe until 20 July, didn't it? And the basic functionality was already implemented by that time. Thus I'd like to know what kind of issue second pipe was meant to address.

**Comment by cshnik** [comment]
Unfortunately, there were different problems with initial implementation: lack of IO errors handling/reporting and some communication problems/deadlocks.

Next changes/improvements were added:
1. Fixed IPC implementation based on a single pipe (as it was in original commit).
2. Additional fix added to address potential deadlock issue.

Please use next [build](https://github.com/ProgerXP/Notepad2e/blob/Process_elevation_%23166/bin/Notepad2e.exe?raw=true) for testing purposes.

**Comment by ProgerXP** [comment]
I have been considering this entire implementation and while I mostly understand how it works, I am very concerned about the complexity of the state machine with all these events and signals going by the pipe. I don't feel it's bullet-proof, it feels like there's too much code to solve the task.

Let's take a step back and think if it can be implemented without pipes at all. I have briefly outlined it before (https://github.com/ProgerXP/Notepad2e/issues/166#issuecomment-402253871). 

To improve on my suggestion:

- we already have several signaling objects at our disposal
  - parent process - exists or not
  - child (elevated) process - exists or not, and its exit code
  - file mapping - exists or not
- in a nutshell, there's 1 common event; when it's first signaled by the first process (non-elevated), second process obtains the name of the target file by sending a message to the main window and creates a file mapping; when it's signaled again, second process flushes and closes the mapping
  - it's easy to recognize when state is out of sync by checking if file mapping exists
- second process (elevated) accepts a random identifier on the command-line and enters a loop which breaks only when its parent process has exited or the event with the random identifier is signaled
  - when parent process exits this process exits too
  - when the event is signaled, send the message and get the file name, then create a mapping
    - meanwhile first process waits until the mapping exists
  - when the event is again signaled, flush and close the mapping and repeat the loop
  - on any error the process exits with a non-0 exit code
- first process (non-elevated) starts the second process (unless it's already running), signals the event, then enters a wait loop, then writes the content to the mapping, signals the event again and reports success if the child process is still running and mapping doesn't exist
  - before each state transition, check if child process exists; if it has exited, then an error has occurred and file buffer is not marked as "saved", and an error is reported to the user
- to simplify the implementation, second process is given also PID of its parent and the window handle since these never change

Don't yet start writing code, let's discuss this - what potential problems you see with this approach?

**Comment by ProgerXP** [comment]
> Actually, no longer available FILEMAPPING can't be treated as a success since another errors may occur, e.g.: unmapping of file view succeeded (FILEMAPPING closed), finalizing of the file (file size adjustment call) failed.

Are you sure? Based on your `FileMapping_Close`, it looks like all errors occur prior to CloseHandle. Therefore we can replace all FileMapping_SaveError with ExitProcess(1).

At least I am assuming that the mapping exists even with no mapped file, i.e. after CreateFile (FileMapping_Open) and CloseHandle. Do you know that the parent process will see the mapping as non-existing 1) before MapViewOfFile 2) or after UnmapViewOfFile?

**Comment by cshnik** [comment]
> Are you sure? Based on your FileMapping_Close, it looks like all errors occur prior to CloseHandle. 
<pre>
FileMapping_Close(FileMapping *pFileMapping, const __int64 size)
...
  if (pFileMapping->lpData && !FileMapping_FlushViewOfFile(pFileMapping))
  {
    <b>FileMapping_SaveError(pFileMapping);</b>
  }
  if (pFileMapping->lpData && !FileMapping_UnmapViewOfFile(pFileMapping))
  {
    <b>FileMapping_SaveError(pFileMapping);</b>
  }
  if (pFileMapping->handle) // <b>file mapping object always closed</b>
  {
    CloseHandle(pFileMapping->handle);
    pFileMapping->handle = NULL;
  }
  if (pFileMapping->file)
  {
    if (size >= 0)
    {
      LONG hiDWORD = HIDWORD(size);
      if (INVALID_SET_FILE_POINTER == SetFilePointer(pFileMapping->file, LODWORD(size), &hiDWORD, FILE_BEGIN))
      {
        <b>FileMapping_SaveError(pFileMapping);</b>
      }
      if (!SetEndOfFile(pFileMapping->file))
      {
         <b>FileMapping_SaveError(pFileMapping);</b>
      }
    }
    CloseHandle(pFileMapping->file);
    pFileMapping->file = NULL;
  }
   ...
</pre>

> Therefore we can replace all FileMapping_SaveError with ExitProcess(1).

It sounds weird. ExitProcess() out of main application procedure is very unhandy and unpleasant. I'm pretty sure this is not an option here.

> Do you know that the parent process will see the mapping as non-existing 1) before MapViewOfFile 2) or after UnmapViewOfFile?

FileMapping class should be used as a black box and for it's intended use. ``FileMapping_Close()`` is used to close mapping and provide error code (in case of error). This specific class should don't be payloaded with foreign/extra functionality which does not directly refer to its original purpose.

> Even if I am correct, there is of course a chance for race condition on ExitProcess: the mapping (process' resources) are closed by the OS slightly before the process ceases to exist. We can combat this by simple Sleep(100) in the parent process after seeing file mapping as closed - if the child process is still running after 100 ms then it's fine, else if it has exited - then it has failed.

1. File mapping should be always closed on ParentProcess_ElevatedFileIO() quit despite of the result (success/failure). There is no need to check if it still exists because there is no reuse logic expected.
2. In this Sleep()-case there is no way to distinguish FileMapping_Close() result from unexpected CHILD-PROCESS termination occurred near the same time (assume file was saved successfully, user/software terminated CHILD-PROCESS in the next 1 ms. PARENT-PROCESS will get faulty file save error).
3. Using of Sleep() in event-based procedure looks **very strange**.
4. The clear and smooth solution is: add "file mapping close"-event to notify PARENT-PROCESS ParentProcess_ElevatedFileIO() has succeeded, e.g.:
<pre>
	Switch (MsgWaitForMultipleObjectsEx (CHILD-PROCESS, FILEMAPPING, <b>EVENT_SUCCESS</b>))
		<b>case EVENT_SUCCESS:
			Exit;</b>
	...
</pre>

<pre>
			Else If EVENT_COUNTER == 1 Then
				If FAILED (CloseFileMapping ()) Then
					ExitApplication (GetLastError ())
				EndIf
				<b>SignalEvent (EVENT_SUCCESS)</b>
				EVENT_COUNTER = 0
			EndIf
</pre>

**Comment by ProgerXP** [comment]
> `FileMapping_Close(FileMapping *pFileMapping, const __int64 size)`


Sorry, I have missed the first CloseHandle and mistakenly taken second CloseHandle after all other operations as last operation on the mapping.

I see what you mean with the second event, of course it's most straightforward approach, but what do you think about these two options:

1. We can try moving `CloseHandle(pFileMapping->handle);` to the bottom of `FileMapping_Close` - since the file is no more mapped to it (`UnmapViewOfFile` remains where it is now) then operations on `pFileMapping->file` should proceed fine in spite of still not closed `pFileMapping->handle`.

```
FlushViewOfFile
UnmapViewOfFile
SetFilePointer
SetEndOfFile
CloseHandle file
CloseHandle mapping
```

If this works, then we can use the mapping as an event signal - if we encounter any error before last CloseHandle, we just exit the process (whether ExitProcess is good or bad is a topic of separate discussion, let's first establish if it's possible to keep the mapping active until the end or not).

2. If the above doesn't work, we can reuse the same event but let the child signal it. This way, parent will enter a nested loop:

```
While (True) Do
	Switch (MsgWaitForMultipleObjectsEx (CHILD-PROCESS, FILEMAPPING))
		case CHILD-PROCESS:
			...
		case FILEMAPPING:
			...
			SignalEvent (EVENT)				// finalize FILEMAPPING <-- problem is here (!)
			** While (...) **
```

Such nested loop would be cleaner I think, but there are two questions: 

- first, the event should not be seen as signaled by the parent (even though it's set as "signaled" by the line above the loop) - maybe we can use PulseEvent which signals and resets the event, that would be seen by the child as signaled (with WaitFor...) but not the parent
- second, if the child can signal the event which it didn't create (if permissions allow it)

**Comment by cshnik** [comment]
> If this works, then we can use the mapping as an event signal - if we encounter any error before last CloseHandle, we just exit the process (whether ExitProcess is good or bad is a topic of separate discussion, let's first establish if it's possible to keep the mapping active until the end or not).

* Delayed object state check after Sleep() is not a good option to get assured operation result (e.g. for file mapping close):
1. Unreliable: any API-call can potentially be retarded due to multiple reasons on the target machine (e.g: low CPU/memory performance, antivirus activity, installing drivers/software, etc) which means actual timeframes required for operation completion may heavily differ.
1. Ineffective and not user-friendly: working thread is completely unresponsive/freezed until Sleep() return.
* This approach does not work since SetEndOfFile()-call returns an error:
``ERROR_USER_MAPPED_FILE: The requested operation cannot be performed on a file with a user-mapped section open``.

> If the above doesn't work, we can reuse the same event but let the child signal it.
> Such nested loop would be cleaner I think, but there are two questions: 
> 
> - first, the event should not be seen as signaled by the parent (even though it's set as "signaled" by the line above the loop) - maybe we can use PulseEvent which signals and resets the event, that would be seen by the child as signaled (with WaitFor...) but not the parent
> - second, if the child can signal the event which it didn't create (if permissions allow it)

1. PulseEvent() is not required since current implementation uses auto-reset events (return of WaitFor-call automatically reset event state).
2. This will not work since there is a chance CHILD-PROCESS will capture this event itself in his main Wait()-loop:
<pre>
While (True) Do
	Switch (WaitForMultipleObjects (PARENT-PROCESS, <b>EVENT</b>))
...
		case <b>EVENT</b>:
...
			Else If EVENT_COUNTER == 1 Then
				If FAILED (CloseFileMapping ()) Then // nested call of SignalEvent (<b>EVENT</b>)
...
</pre>
This was the reason of initial switching to 2 pipes implementation (every pipe had own assigned event) and excluded possibility of accidental wrong event handling.

**Comment by ProgerXP** [comment]
> Delayed object state check after Sleep() is not a good option to get assured operation result (e.g. for file mapping close):

Sleeping was a quick and dirty suggestion. It's not as important as discovering if we can use the mapping as an event at all.

> This approach does not work since SetEndOfFile()-call returns an error

In other words, even though you have removed that file's mapping with UnmapViewOfFile, SetEndOfFile still sees the file as mapped? And SetEndOfFile only agrees to operate when the mapping is closed entirely (CloseHandle)?

What if you map another file onto the same mapping? I wonder if it will be now attached to two files at the same time:

```
UnmapViewOfFile (file 1)
MapViewOfFile (file 2)
SetEndOfFile (file 1)
// result?
```

> This was the reason of initial switching to 2 pipes implementation (every pipe had own assigned event) and excluded possibility of accidental wrong event handling.

This resulted in a very tangled logic as well as reading/writing buffers **combined** with signaling events. At least we should try to find an alternative.

**Comment by cshnik** [comment]
Filemapping is tied to concrete file:
<pre>
      pFileMapping->handle = CreateFileMapping(pFileMapping-><b>file</b>, NULL, PAGE_READWRITE,
                                               HIDWORD(size), LODWORD(size), pFileMapping->name);
</pre>
There is no way to `map another file onto the same mapping`.

> In other words, even though you have removed that file's mapping with UnmapViewOfFile, SetEndOfFile still sees the file as mapped? And SetEndOfFile only agrees to operate when the mapping is closed entirely (CloseHandle)?

UnmapViewOfFile() only releases specific data range (view) accessible via file mapping object. The file remains mapped until file mapping object closed.

> This resulted in a very tangled logic as well as reading/writing buffers combined with signaling events. 
Complete list of events required for reliable IPC implementation is:
<pre>
EVENT_FILE_MAPPING_CREATE_REQUEST	(requested by PARENT-PROCESS)
EVENT_FILE_MAPPING_CREATED		(responded by CHILD-PROCESS)
EVENT_FILE_MAPPING_CLOSE_REQUEST	(requested by PARENT-PROCESS)
EVENT_FILE_MAPPING_CLOSED		(responded by CHILD-PROCESS)
</pre>
All of them can be implemented as a part of FileMapping()-class.

**Comment by cshnik** [comment]
BTW Current implementation of file mapping uses internal regular event object to signal its ready state:
<pre>
struct TFileMapping
{
  HANDLE file;
  HANDLE handle;
  WCHAR name[MAX_PATH];
  <b>Event event;</b>
...
};
</pre>
<pre>
BOOL FileMapping_Set(FileMapping *pFileMapping)
{
...
  return Event_Set(&pFileMapping-><b>event</b>);
}
</pre>
It looks like extending of FileMapping()-class (implementing of additional events) is not a bad decision.

**Comment by ProgerXP** [comment]
> File mapping is tied to concrete file:

My bad, I have missed again that the structure holds 2 separate handlers.

Okay, it seems there's no way around a second event for signaling, it is indeed the most straightforward solution I can think of. A Sleep call would be even better to make things look synchronous but you are correct that it's not reliable.
 
> Complete list of events required for reliable IPC implementation is:

> BTW Current implementation of file mapping uses internal regular event object to signal its ready state:

Okay, so your idea is to create 4 events to simplify both parent and child loops? So that:

- it would be a single loop (with WaitFor...) with 4 possible break points (plus 1 - the dependent process has exited)
- no pipes, only 1 extra message for retrieving the file name
- error codes are propagated from child to parent via process exit code

If so, then please write a more detailed pseudocode example for final confirmation.

**Comment by ProgerXP** [comment]
Notes on your original pseudocode:

> If ElevatedMode Is Off Then

You mean "when elevated mode cannot be activated (WinXP)"?

> ExitApplication (CHILD-PROCESS, GetLastError ()) // IO operation failed, force CHILD-PROCESS quit with required code

Good idea but can a non-privileged process really terminate a privileged one even if it has spawned it?

> Elevated process will not start on "Elevate" menu item click since CHILD-PROCESS executed from ParentProcess_ElevatedFileIO() only.

Why so? Even if the child is launched, it simply enters the loop waiting for the signal - if we only launch the process without signaling the event (as from File > Elevate), then the child will wait until the user uses Save.

---

In general do you think it will be simpler than the original implementation with pipe(s)?

**Comment by cshnik** [comment]
>> If ElevatedMode Is Off Then

> You mean "when elevated mode cannot be activated (WinXP)"?

Next conditions must be met to allow execution of elevation IPC:
1. Current OS version is higher than WinXP
2. Current instance is not elevated.
3. Elevation mode is enabled.

> Good idea but can a non-privileged process really terminate a privileged one even if it has spawned it?

I was able to test required behavior in my local tests. This scenario works as expected: PARENT-PROCESS succeeded to start new elevated instance (CHILD-PROCESS), execute ``TerminateProcess(CHILD-PROCESS, exit-code)`` and retrieve ``exit-code`` value with subsequent ``GetExitCodeProcess()``-call.

>>Elevated process will not start on "Elevate" menu item click since CHILD-PROCESS executed from ParentProcess_ElevatedFileIO() only.

> Why so? Even if the child is launched, it simply enters the loop waiting for the signal - if we only launch the process without signaling the event (as from File > Elevate), then the child will wait until the user uses Save.

Ok. This means there should be additional code in the menu item handler to start/terminate CHILD-PROCESS on item state change.

Please consider next updated code for IPC.
Changes in **Main Window Procedure**:

```
switch (msg.message)
...
case WM_GET_FILEIO_FILENAME:
	ProvideFileIODestinationFilename (msg);
	break;
...
	case IDM_FILE_ELEVATE:	// elevation menu item handler
		If ElevationMode is Enabled Then
			RunApplication (CHILD-PROCESS)
		ElseIf
			ExitApplication (CHILD-PROCESS)
		EndIf
...
```

**ParentProcess_ElevatedFileIO():**
```
If ElevatedMode Is Off Then
	RETURN FAILURE
EndIf
If CHILD-PROCESS Is Not Running Then
	RunApplication (CHILD-PROCESS)
	If CHILD-PROCESS Is Not Running Then
		HandleError (CHILD_PROCESS_NOT_EXECUTED)
		RETURN FAILURE
	EndIf
EndIf

FILEMAPPING.TryCreate ()				// internally signal the FILEMAPPING.EventCreateRequest

While (True) Do
	Switch (MsgWaitForMultipleObjectsEx (CHILD-PROCESS, FILEMAPPING.EventCreated, FILEMAPPING.EventClosed))
		case CHILD-PROCESS:
			HandleError (GetExitCode (CHILD-PROCESS))			// treat application return code as IO operation error code
			RETURN FAILURE
		case FILEMAPPING.EventCreated:
			If SUCCEEDED (FILEMAPPING.Open ()) And SUCCEEDED (FileIO (FILEMAPPING)) Then
				FILEMAPPING.TryClose ()	// internally signal the FILEMAPPING.EventCloseRequest
			ElseIf
				ExitApplication (CHILD-PROCESS, GetLastError ())	// IO operation failed, force CHILD-PROCESS to quit with specific code
			EndIf
		case FILEMAPPING.EventClosed:
			RETURN SUCCESS			// elevated IO operations completed, quit
		case MSG_WAIT_OBJECT:			// process WM_GET_FILEIO_FILENAME in **Main Window Procedure**
			ProcessPendingMessages ()
	End Switch
Next While
```


**ChildProcess_FileIOHandler():**
```
While (True) Do
	Switch (WaitForMultipleObjects (PARENT-PROCESS, FILEMAPPING.EventCreateRequest, FILEMAPPING.EventCloseRequest))
		case PARENT-PROCESS:
			ExitApplication (0)
		case FILEMAPPING.EventCreateRequest:
			If SUCCEEDED (SendMessage_RetrieveDestinationFilename ()) Then
				if FAILED (FILEMAPPING.Create ()) Then		// internally signal  the FILEMAPPING.EventCreated in case of success
					ExitApplication (GetLastError ())
				EndIf
			ElseIf
				ExitApplication (GetLastError ())
			EndIf
		case FILEMAPPING.EventCloseRequest:
			If FAILED (FILEMAPPING.Close ()) Then			// internally signal the FILEMAPPING.EventClosed in case of success
				ExitApplication (GetLastError ())
			EndIf
	End Switch
Next While
```

**Comment by ProgerXP** [comment]
Okay, this looks way cleaner and more straightforward than the solution using the pipe(s). You can code this. If you can make some rough draft version, I can also review the real code.

Do you think it will be indeed easier and more reliable?

**Comment by cshnik** [comment]
> Do you think it will be indeed easier and more reliable?

This scheme is definitely easier for reading/understanding.

Please check [test build](https://github.com/ProgerXP/Notepad2e/raw/8b2bfd3a127f7e4e174898e0277b6cf8f1d91a5e/bin/Notepad2e.exe) for current implementation with ExitProcess().

Here are some points:
1. Auxiliary filemapping instance (``fileMappingIPCData``) was used to transfer data between Parent to Child processes (shared memory object, not linked to a file on disk) instead of proposed SendMessage()-call. The reasons are:
- There is no easy way to share strings/memory blocks between processes using Windows messaging mechnism.
- WM_COPYDATA can't be used to request data from CHILD process. Any other possible implementations are awkward.
- Implementation of the alternative data sharing mechanism is not required when we reuse filemapping.
- Shared memory via file mapping is the best way to implement IPC.
- Reusing of the existing code (filemapping class) is a good practice we should follow.

Thus, IPC scheme has changed:
+ Request ``fileMappingIPCData`` creation
+ Transfer filename and size to CHILD process
+ Request ``fileMapping`` creation
+ Transfer data to destination file
2. CHILD process is restarted (with elevation screen popup) after every error (e.g. access denied issue due to destination file lock from foreign application).
3. Error and object handling must be carefully checked (TODO).

**Comment by ProgerXP** [comment]
The code looks fine, much better than the previous attempt. You can now finish it. Some comments below.

> const int iMaxCmdLength = lstrlen(lpCmdLine) + 25;
> LPWSTR lpCmdLineNew = n2e_Alloc(sizeof(WCHAR) * iMaxCmdLength);
> lstrcpyn(lpCmdLineNew, lpCmdLine, iMaxCmdLength);
> lstrcat(lpCmdLineNew, L" /" IPCID_PARAM);
> lstrcat(lpCmdLineNew, toString(GetCurrentProcessId()));
> lstrcat(lpCmdLineNew, L",");
> lstrcat(lpCmdLineNew, toString(dwIPCID));

I believe this won't work with all combinations of parameters. In particular, if `/Z` is used then the IPC argument will be ignored. Also, it may not work correctly with environment variables since two processes are running with different environment (regular user and admin user). Working directory sometimes may be an issue too.

Why do we need to pass the original command line anyway? I think constructing it manually like `c:\absolute\file.txt /IPCID=%d,%d` is enough. We don't even need ExtractFirstArgument because the path to EXE is given separately via `si`.

> LPCWSTR getObjectName2(LPCWSTR lpPrefix, LPCWSTR lpPostfix)

Do we need this?

> BOOL n2e_InitializeIPC(const DWORD idIPC, const BOOL bClientMode)
> BOOL n2e_InitIPC(const BOOL bResetIPC)

These 2 functions basically have the same name. Rename one of them please, or delete (below).

> BOOL n2e_InitIPC(const BOOL bResetIPC)
> const DWORD ipc = (!bResetIPC && (i == 1)) ? dwIPCID : GenerateIPCID();

We don't need bResetIPC anymore, right? This removes the need for this assignment.

Moreover, I think we don't need this function and a loop at all. Chances of failing are low, and we can remove them entirely by using current process ID instead of rand() - Windows maintains unique PIDs therefore there is no chance that dwIPCID will be ever duplicated if we use parent PID as IPCID.

This also means we don't need to pass 2 numbers via /IPCID, only 1 is enough.

> HBITMAP ConvertIconToBitmap(const HICON hIcon, const int cx, const int cy)
> void n2e_SetUACIcon(HMENU hMenu, UINT nItem)

Shouldn't these be in Shell32Helper? They are generic and will reduce the amount of code in that file.

**Comment by cshnik** [comment]
> > const int iMaxCmdLength = lstrlen(lpCmdLine) + 25;
> > LPWSTR lpCmdLineNew = n2e_Alloc(sizeof(WCHAR) * iMaxCmdLength);
> > lstrcpyn(lpCmdLineNew, lpCmdLine, iMaxCmdLength);
> > lstrcat(lpCmdLineNew, L" /" IPCID_PARAM);
> > lstrcat(lpCmdLineNew, toString(GetCurrentProcessId()));
> > lstrcat(lpCmdLineNew, L",");
> > lstrcat(lpCmdLineNew, toString(dwIPCID));
> 
> I believe this won't work with all combinations of parameters. In particular, if `/Z` is used then the IPC argument will be ignored. Also, it may not work correctly with environment variables since two processes are running with different environment (regular user and admin user). Working directory sometimes may be an issue too.
Why do we need to pass the original command line anyway? I think constructing it manually like `c:\absolute\file.txt /IPCID=%d,%d` is enough. We don't even need ExtractFirstArgument because the path to EXE is given separately via `si`.

Most code of ``n2e_RunElevatedInstance()`` was copied from original ``RelaunchElevated()`` which contains exactly the same ExtractFirstArgument()-call.
Original method ``RelaunchElevated`` is called when running non-elevated application with ``/U`` command line switch.
Please confirm if we should change this logic when running elevated instance for IPC.

> > LPCWSTR getObjectName2(LPCWSTR lpPrefix, LPCWSTR lpPostfix)
> 
> Do we need this?
> 
> > BOOL n2e_InitializeIPC(const DWORD idIPC, const BOOL bClientMode)
> > BOOL n2e_InitIPC(const BOOL bResetIPC)
> 
> These 2 functions basically have the same name. Rename one of them please, or delete (below).

Fixed.

> This also means we don't need to pass 2 numbers via /IPCID, only 1 is enough.

Fixed.



> > HBITMAP ConvertIconToBitmap(const HICON hIcon, const int cx, const int cy)
> > void n2e_SetUACIcon(HMENU hMenu, UINT nItem)
> 
> Shouldn't these be in Shell32Helper? They are generic and will reduce the amount of code in that file.

Actually no, Shell32Helper was added as a proxy module for shell32 system library. It allow to compile application for Windows XP and use extended shell functionality when running on modern OSes.
This methods were moved to MainWndHelper module which already contains similar specific stuff.

Please check next [test build](https://github.com/ProgerXP/Notepad2e/blob/Process_elevation_%23166/bin/Notepad2e.exe?raw=true).

**Comment by ProgerXP** [comment]
> Most code of n2e_RunElevatedInstance() was copied from original RelaunchElevated() which contains exactly the same ExtractFirstArgument()-call. 
> Please confirm if we should change this logic when running elevated instance for IPC.

For n2e_RunElevatedInstance(), the logic should be changed. It's different from /U because in the latter case we need to re-launch the program with exactly the same behaviour - so using the same command-line. With our IPC client/slave we want the process to do only one specific thing and we know that no other command-line parameters are necessary for this operation. Therefore we should not append to the original arguments as this could provoke unwanted side effects.

> void CloseProcessHandle
> LPCWSTR formatObjectName

Is there any reason you're naming some functions capitalized while others lowercased? I got an impression Notepad2 is using PascalCase, so should we.

> void n2e_ChildProcess_Exit(const DWORD dwExitCode, HANDLE hParentProcess)

```
{
  CloseProcessHandle(&hParentProcess);
  n2e_FinalizeIPC();
  ExitProcess(dwExitCode);
}
```

Is CloseProcessHandle necessary? Isn't Windows automatically closing all handles when the process is terminated, which happens inevitably by the following ExitProcess?

> n2e_ChildProcess_FileIOHandler

```
        n2e_ChildProcess_Exit(0, hParentProcess);
        return;
```

Why `return;`? It looks as if there was some special meaning compared to `break;` in other places. I suggest we use `break;` here as well - this has the same effect given the preceding `n2e_ChildProcess_Exit` call.

> n2e_ParentProcess_ElevatedFileIO

I think that bResult and bContinue make the logic actually harder to follow. 

```
  BOOL bResult = FALSE;
  ...

  if (!n2e_IsElevatedModeEnabled())
  {
    return bResult;
  }
```

Just `return FALSE;`?

Next, in the loop, bResult is only set once, and bContinue is set there too so this is also `return TRUE;`. In the second place with bContinue (WAIT_OBJECT_0), this is again simply `return FALSE;`. There's no finalization logic in this loop.

With Result/Continue I had to check this beforehand because it wasn't clear if, for example, WAIT_OBJECT_0 could ever return TRUE. 

Then the loop can be made infinite (like you did in n2e_ChildProcess_FileIOHandler) which would simplify the understanding.

**Comment by cshnik** [comment]
> Therefore we should not append to the original arguments as this could provoke unwanted side effects.

Fixed.

> Is there any reason you're naming some functions capitalized while others lowercased? I got an impression Notepad2 is using PascalCase, so should we.

Fixed.

> Is CloseProcessHandle necessary? Isn't Windows automatically closing all handles when the process is terminated, which happens inevitably by the following ExitProcess?

Actually no, neither CloseProcessHandle  nor n2e_FinalizeIPC() are required since `ExitProcess` [documentation](https://docs.microsoft.com/en-us/windows/desktop/api/processthreadsapi/nf-processthreadsapi-exitprocess) directly states:
> 6. All of the object handles opened by the process are closed.

Please consider next points:
1. ``n2e_ChildProcess_Exit`` is an exit point for IPC.
2. This procedure was implemented to make the only place of calling an exceptional API method `ExitProcess`.
3. It is a good practice to manually care of the resources release. And it costs nothing to add these two extra finalization calls.

> > n2e_ChildProcess_FileIOHandler
> 
> ```
>         n2e_ChildProcess_Exit(0, hParentProcess);
>         return;
> ```
> Why `return;`? It looks as if there was some special meaning compared to `break;` in other places. I suggest we use `break;` here as well - this has the same effect given the preceding `n2e_ChildProcess_Exit` call.

Original meaning for `return` was: parent process quit caused child process to stop IPC IO handler work. This is a higher level logic than implemented way of procedure quit and/or error handling (currently used nested call to `ExitProcess`). Please confirm if you insist on a more uniform but less meaning `break;`.

> I think that bResult and bContinue make the logic actually harder to follow.

Fixed. `n2e_ParentProcess_ElevatedFileIO` was switched to an infinite loop scheme.

Please check [test build](https://github.com/ProgerXP/Notepad2e/blob/d39298af66531d304ac92da86c2c013db11ceaa1/bin/Notepad2e.exe?raw=true).

**Comment by cshnik** [comment]
> Isn't it simpler and more reliable to use GetModuleFileName that's guaranteed to return a real, full path? No need for memory management too.

Fixed.

> Is this implementation fully tested and potentially ready for merge?

I see no shortcomings in the current implementation. [Test build](https://github.com/ProgerXP/Notepad2e/blob/ea3c190738ed5a1e4ae0dfea30afa2f7efc95d44/bin/Notepad2e.exe?raw=true).

**Comment by ProgerXP** [comment]
```
  static BOOL bIsInitialized = FALSE;
  if (!bIsInitialized)
  {
    bIsInitialized = TRUE;
    STARTUPINFO si = { 0 };
    GetStartupInfo(&si);
    bIsParentIPC = !((si.dwFlags & STARTF_USESHOWWINDOW) && (si.wShowWindow == SW_HIDE));
  }
```

What is this new block in `n2e_InitializeIPC` meant to accomplish?

---

> I see no shortcomings in the current implementation. 

First thing I try is do Elevate and cancel the UAC prompt and I get no error messages and the Elevate menu item becomes checked. Are you sure it was properly tested?

**Comment by ProgerXP** [comment]
> I think the new file size is miscalculated. 

I figured that it's happening when document encoding is one of the `Unicode ... BOM` types such as UTF-16 LE/BE BOM.

I see in the code you have detection of BOM and buffer size correction, but why it doesn't work?

**Comment by ProgerXP** [comment]
> The block was implemented to get rid of external parameter (is parent/child mode) to rely on the initial window visibility state (child instance is always hidden).

It seems to me that relying on such factors is obscure and risky. What was wrong with passing the parameter explicitly? After all, given the command line parameter the process always knows if it's a client or a server, I don't see why we should check for window visibility instead of using this definite factor (CLI argument).

> There was a bug related to missing text buffer conversion 

Can you remind me why we have to provide exact file size to the elevated instance and why it can't just truncate the file after the master has written all data via the file mapping?

Pre-calculating file size seems to be error-prone even though we've fixed all issues we could find.

**Comment by cshnik** [comment]
> > The block was implemented to get rid of external parameter (is parent/child mode) to rely on the initial window visibility state (child instance is always hidden).
> 
> It seems to me that relying on such factors is obscure and risky. What was wrong with passing the parameter explicitly? After all, given the command line parameter the process always knows if it's a client or a server, I don't see why we should check for window visibility instead of using this definite factor (CLI argument).

The point was to isolate implementation features:
1. Current implementation uses the only argument to initialize IPC: specific communication ID (passed to ``n2e_InitializeIPC``). The only ways/places to initialize IPC are:
   * Enable elevation from system menu (ID will be generated in the feature's module (FM) ProcessElevationUtils.c)
   * Startup of elevated instance from FM (ID will be passed/extracted via the command line).
2. Child/parent mode is an internal option of the feature's code base. Making such an argument public will disclose internal structure/design and potentially may cause incorrect function use.

Please confirm if you would like to rollback this change.

> > There was a bug related to missing text buffer conversion
> 
> Can you remind me why we have to provide exact file size to the elevated instance and why it can't just truncate the file after the master has written all data via the file mapping?
> 
> Pre-calculating file size seems to be error-prone even though we've fixed all issues we could find.

Please consider the following points:
1. File mapping creation requires file size, [Creating a File Mapping Object](https://docs.microsoft.com/en-us/windows/desktop/memory/creating-a-file-mapping-object):
> File Mapping Size
> The size of the file mapping object is independent of the size of the file being mapped. 
> ...
> Doing this creates a file mapping object that is exactly the same size as the file. Otherwise, you must calculate or estimate the size of the finished file because file mapping objects are static in size; once created, their size cannot be increased or decreased. 
2. We should always call ``SetEndOfFile`` to properly commit file size. This is required to truncate target file to delete the data which were out of the current file mapping (i.e working with file mapping smaller than current file size does not affect out of bounds data and actual file size).

**Comment by ProgerXP** [comment]
> Child/parent mode is an internal option of the feature's code base. Making such an argument public will disclose internal structure/design and potentially may cause incorrect function use.

It sounds correct but nonetheless, I don't like the way you're determining if it's server or client in n2e_InitializeIPC. Using startup info, whether the window should be visible or not, is not future-proof. If for any reason we make the window visible, it will break. It's a concealed dependency - will you even remember that you should not ever make window visible or process elevation will fail? I wouldn't have any clue why one is connected to another.

From this point of view, sharing a flag, even if it's an implementation detail, is better, because it's an explicit dependency/statement.

> File mapping creation requires file size, Creating a File Mapping Object:

I see.

**Comment by cshnik** [comment]
> From this point of view, sharing a flag, even if it's an implementation detail, is better, because it's an explicit dependency/statement.

Done. There was a rollback commit to directly provide if server or client mode is required. Please confirm if we are ready for merge into master.

**Comment by ProgerXP** [comment]
I have missed the rollback commit.

> Please confirm if we are ready for merge into master.

Yes, please merge. Also see if the other branch (`Boost...`) is necessary, I think it can be deleted as well.

**Comment by cshnik** [comment]
> Yes, please merge. Also see if the other branch (`Boost...`) is necessary, I think it can be deleted as well.

Done.
1. Useless branch ``Boost_regex_and_Cyrillic_#162`` was deleted.
2. Branch ``Process_elevation_#166`` was merged into ``master``. Please confirm if it should be also deleted.

**Comment by ProgerXP** [comment]
I have a failing test now:

```
CStringToHex.TestHex_FileSamples
Assert failed. Decoding, Unicode encoding,  source: TestFile1__src_UTF8.txt
```

> Branch Process_elevation_#166 was merged into master. Please confirm if it should be also deleted.

If you have merged full change history and nothing will be lost then yes.

**Comment by cshnik** [comment]
> I have a failing test now

In my local tests there are no issues with tests using VS2017 in Debug/x86, Debug ICU/x86 configurations. Please provide details on your VS version and used configuration type if the issue still exists.

> Also, we have a clash: make &Elevate Ele&vate.

Done.

> If you have merged full change history and nothing will be lost then yes.

I offer to postpone this branch deletion until we agreed/make sure this major feature works as expected.

**Comment by ProgerXP** [comment]
> I offer to postpone this branch deletion until we agreed/make sure this major feature works as expected.

I haven't seen any problems with this implementation so it's safe to delete.

**Comment by cshnik** [comment]
> I haven't seen any problems with this implementation so it's safe to delete.

Done.

**Comment by ProgerXP** [comment]
For quite some time now attempt to
